src/libraries/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Android/ChainPal.cs

```diff
@@ -35,7 +38,319 @@ public static bool ReleaseSafeX509ChainHandle(IntPtr handle)
+                            extraCerts.Add(certHandle);
+                        }
+                    }
+                }
+
+                IntPtr[] extraArray = extraCerts.ToArray();
+                _chainContext = Interop.AndroidCrypto.X509ChainCreateContext(
+                    ((AndroidCertificatePal)cert).SafeHandle,
+                    extraArray,
+                    extraArray.Length);
+
+                if (useCustomRootTrust)
+                {
+                    // Android does not support an empty set of trust anchors
+                    if (customTrustCerts.Count == 0)
+                    {
+                        throw new PlatformNotSupportedException(SR.Chain_EmptyCustomTrustNotSupported);
+                    }
+
+                    IntPtr[] customTrustArray = customTrustCerts.ToArray();
+                    int res = Interop.AndroidCrypto.X509ChainSetCustomTrustStore(_chainContext, customTrustArray, customTrustArray.Length);
+                    if (res != 1)
+                    {
+                        throw new CryptographicException();
+                    }
+                }
+            }
+
+            internal void Evaluate(
+                DateTime verificationTime,
+                OidCollection applicationPolicy,
+                OidCollection certificatePolicy,
+                X509RevocationMode revocationMode,
+                X509RevocationFlag revocationFlag)
+            {
+                Debug.Assert(_chainContext != null);
+
+                long timeInMsFromUnixEpoch = new DateTimeOffset(verificationTime).ToUnixTimeMilliseconds();
+                _isValid = Interop.AndroidCrypto.X509ChainBuild(_chainContext, timeInMsFromUnixEpoch);
+                if (!_isValid)
+                {
+                    // Android always validates name, time, signature, and trusted root.
+                    // There is no way bypass that validation and build a path.
+                    ChainElements = Array.Empty<X509ChainElement>();
+
+                    Interop.AndroidCrypto.ValidationError[] errors = Interop.AndroidCrypto.X509ChainGetErrors(_chainContext);
+                    var chainStatus = new X509ChainStatus[errors.Length];
+                    for (int i = 0; i < errors.Length; i++)
+                    {
+                        Interop.AndroidCrypto.ValidationError error = errors[i];
+                        chainStatus[i] = ValidationErrorToChainStatus(error);
+                        Marshal.FreeHGlobal(error.Message);
+                    }
+
+                    ChainStatus = chainStatus;
+                    return;
+                }
+
+                if (revocationMode != X509RevocationMode.NoCheck)
+                {
+                    if (revocationFlag == X509RevocationFlag.EntireChain)
+                    {
+                        throw new NotImplementedException($"{nameof(Evaluate)} (X509RevocationFlag.{revocationFlag})");
```

Comment:
This is a weird error to emit to callers.

Probably just treat it as EntireChainIgnoreRoot, as that's the best-fit mapping.  In practice, roots don't get revoked (they have to revoke themselves, which is weird).

Reply:
Would you rather treat it as EntireChainIgnoreRoot or actually throw PNSE?

Reply:
I'd rather treat it as EntireChainIgnoreRoot, since all that anyone will do is "if android, use EntireChainIgnoreRoot instead of EntireChain".  We can do that for them, and there's not much semantic difference.

If it was "EndCertificateOnly is the only mode we support" there's a major risk there, and we'd either need to fail, or work around the problem by "OK, so that chain passed... pop up a level and try again...".  (Besides, maybe we could get EntireChain by doing "fine, we'll build the chain again just for the root... happy?" but we'd want to observe that they don't just ignore revocation checks on self-issued certificates before committing that code)